morgoth/base/utils.py

```diff
@@ -0,0 +1,70 @@
+import requests
+
+from datetime import datetime
+
+from django.conf import settings
+
+
+class BalrogAPI(object):
+    def __init__(self, auth=None):
+        """ Creates an object which wraps the Balrog REST API."""
+        self.auth = auth
+        self.session = requests.session()
+        self.csrf_token = None
+
+    def _is_csrf_token_expired(self):
+        if not self.csrf_token:
+            return True
+
+        expiry = self.csrf_token.split('##')[0]
+        if expiry <= datetime.now().strftime('%Y%m%d%H%M%S'):
+            return True
+
+        return False
+
+    def _make_request(self, url, data, method):
+        headers = {'Accept-Encoding': 'application/json',
+                   'Accept': 'application/json'}
+
+        url = '%s%s' % (settings.BALROG_API_BASE_URL, url)
+
+        req = self.session.request(method=method, url=url, data=data,
+                                   timeout=settings.BALROG_API_REQUEST_TIMEOUT,
+                                   verify=False, auth=self.auth, headers=headers)
+
+        req.raise_for_status()
+        return req
+
+    def request(self, url, prerequest_url=None, url_vars=None, data=None, method='GET'):
+        url = url % url_vars
+
+        data['csrf_token'] = self.csrf_token
+
+        if prerequest_url is not None:
+            prerequest_url = prerequest_url % url_vars
+        else:
+            prerequest_url = url
+
+        # If we're modifying things, do a GET first to get a CSRF token and maybe a data_version.
+        if method not in ('GET', 'HEAD'):
+            # Use the URL of the resource we're going to modify first, because we'll need a CSRF
+            # token, and maybe it's data version.
+            try:
+                res = self._make_request(prerequest_url, None, 'HEAD')
+
+                # If a data_version was specified we shouldn't overwrite it because the caller may
+                # be acting on a modified version of a specific older version of the data.
+                if 'data_version' not in data:
+                    data['data_version'] = res.headers['X-Data-Version']
```

Comment:
This behavior seems odd. When would you be modifying a pre-existing object, but not know what it's `data_version` is? Done this way, it seems really easy to break any mid-air collision Balrog has.